pkg/cloudprovider/gcp.go

```diff
@@ -1,47 +1,205 @@
+	cloudSecret.Data = map[string][]byte{
+		"cloud": secret.Data[GcpCredentials],
+	}
 }
 
 func (p *GCPProvider) UpdateRegistrySecret(secret, registrySecret *kapi.Secret) {
+	registrySecret.Data = map[string][]byte{
+		"cloud": secret.Data[GcpCredentials],
+	}
 }
 
-func (p *GCPProvider) UpdateRegistryDC(deploymentconfig *appsv1.DeploymentConfig, name, dirName string) {
+func (p *GCPProvider) UpdateRegistryDC(dc *appsv1.DeploymentConfig, name, dirName string) {
+	dc.Spec.Template.Spec.Containers[0].Env = []kapi.EnvVar{
+		{
+			Name:  "REGISTRY_STORAGE",
+			Value: "gcs",
+		},
+		{
+			Name:  "REGISTRY_STORAGE_GCS_BUCKET",
+			Value: p.Bucket,
+		},
+		{
+			Name:  "REGISTRY_STORAGE_GCS_ROOTDIRECTORY",
+			Value: "/" + dirName,
+		},
+		{
+			Name:  "REGISTRY_STORAGE_GCS_KEYFILE",
+			Value: "/credentials/cloud",
+		},
+	}
+	dc.Spec.Template.Spec.Containers[0].VolumeMounts = []kapi.VolumeMount{
+		{
+			Name:      "cloud-credentials",
+			MountPath: "/credentials",
+		},
+	}
+	dc.Spec.Template.Spec.Volumes = []kapi.Volume{
+		{
+			Name: "cloud-credentials",
+			VolumeSource: kapi.VolumeSource{
+				Secret: &kapi.SecretVolumeSource{
+					SecretName: name,
+				},
+			},
+		},
+	}
 }
 
 func (p *GCPProvider) Validate(secret *kapi.Secret) []string {
 	fields := []string{}
 
+	if secret != nil {
+		keySet := []string{
+			GcpCredentials,
+		}
+		for _, k := range keySet {
+			p, _ := secret.Data[k]
+			if p == nil || len(p) == 0 {
+				fields = append(fields, "Secret(content)")
+				break
+			}
+		}
+	}
+
 	switch p.Role {
 	case BackupStorage:
+		if p.Bucket == "" {
+			fields = append(fields, "Bucket")
+		}
 	case VolumeSnapshot:
 	}
 
 	return fields
 }
 
 func (p *GCPProvider) Test(secret *kapi.Secret) error {
+	var err error
 	if secret == nil {
 		return nil
 	}
 
+	switch p.Role {
+	case BackupStorage:
+		key, _ := uuid.NewUUID()
+		test := GcsTest{
+			key:    key.String(),
+			bucket: p.Bucket,
+			secret: secret,
+		}
+		err = test.Run()
+		if err != nil {
+			return err
+		}
+	case VolumeSnapshot:
+		// TBD
+	}
+
+	return nil
+}
+
+type GcsTest struct {
+	bucket string
+	secret *kapi.Secret
+	key    string
+}
+
+func (r *GcsTest) Run() error {
+	client, err := r.newClient()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	err = r.upload(client)
+	if err != nil {
+		return err
+	}
+	defer r.delete(client)
+	err = r.download(client)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
+
+func (r *GcsTest) newClient() (*storage.Client, error) {
+	client, err := storage.NewClient(
+		context.Background(),
+		option.WithScopes(storage.ScopeReadWrite),
+		option.WithCredentialsJSON(r.secret.Data[GcpCredentials]))
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (r *GcsTest) upload(client *storage.Client) error {
+	bucket := client.Bucket(r.bucket)
+	object := bucket.Object(r.key)
+	writer := object.NewWriter(context.Background())
+	_, err := writer.Write([]byte{0})
+	writer.Close()
```

Comment:
Need error checking on `Close` here. Connection checking is currently failing with 403 errors in my cluster, but the error is only being reported on the attempted read, even though the attempted write should be failing too. In looking at the docs for the vendor code here, the following comes from the `Write` function:
```
// Write appends to w. It implements the io.Writer interface.
//
// Since writes happen asynchronously, Write may return a nil
// error even though the write failed (or will fail). Always
// use the error returned from Writer.Close to determine if
// the upload was successful.
```
I think the right thing to do here is to store the return error for Close in a different local var and if either err or <whatever you call the Close return val> is non-nil, then return the non-nil error -- this way we still at least attempt to close the writer even if Write returned an error.

Reply:
Fixed.